src/compiler/checker.ts

```diff
@@ -27181,7 +27181,17 @@ namespace ts {
             // and that call signature is non-generic, return statements are contextually typed by the return type of the signature
             const signature = getContextualSignatureForFunctionLikeDeclaration(functionDecl as FunctionExpression);
             if (signature && !isResolvingReturnTypeOfSignature(signature)) {
-                return getReturnTypeOfSignature(signature);
+                const functionFlags = getFunctionFlags(functionDecl);
+
+                return filterType(getReturnTypeOfSignature(signature), returnType => {
+                    if (functionFlags & FunctionFlags.Generator) {
+                        return checkGeneratorInstantiationAssignabilityToReturnType(returnType, functionFlags, /*errorNode*/ undefined);
+                    }
+                    if (functionFlags & FunctionFlags.Async) {
+                        return !!getAwaitedTypeOfPromise(returnType);

```

Comment:
This is a `isPromise` check based on the existing one in the codebase:
https://github.dev/microsoft/TypeScript/blob/98c19cbbbe83c2ae3c89a4e08317a4b9ccbcb206/src/compiler/checker.ts#L38691